src/harness/testcases/WINNF_FT_S_SIQ_testcase.py

```diff
@@ -841,6 +842,235 @@ def test_WINNF_FT_S_SIQ_11(self):
+        ppa['ppaRecord']['ppaInfo']['cbsdReferenceId'].append(cbsd_ids[device_index])
+      # Inject PPA into SAS UUT
+      zone_id = self._sas_admin.InjectZoneData({"record": ppa['ppaRecord']})
+ 
+    # 5. Trigger CPAS activity if GWPZ > 0 or PPA > 0
+    if (len(config['gwpzRecords']) > 0 or len(config['ppaRecords']) > 0):
+      self.TriggerDailyActivitiesImmediatelyAndWaitUntilComplete() 
+
+    # 6. Send N2 spectrum inquiry requests (one per registered CBSD)
+    spectrum_inquiry_list = config['spectrumInquiryRequests']
+    addCbsdIdsToRequests(cbsd_ids, spectrum_inquiry_list)
+ 
+    spectrum_inquiry_request = {'spectrumInquiryRequest': spectrum_inquiry_list}
+    spectrum_inquiry_response = self._sas.SpectrumInquiry(spectrum_inquiry_request)['spectrumInquiryResponse']
+
+    # Check the response contains N2 SpectrumInquiryResponse objects
+    self.assertEqual(len(spectrum_inquiry_response),len(config['spectrumInquiryRequests']))
+
+    for index,response in enumerate(spectrum_inquiry_response):
+      # If the corresponding request contained a valid cbsdId, the
+      # response shall contain the same cbsdId.
+      if 'cbsdId' in spectrum_inquiry_list[index]:
+        if spectrum_inquiry_list[index]['cbsdId'] in cbsd_ids:
+          self.assertEqual(response['cbsdId'], spectrum_inquiry_list[index]['cbsdId'])
+        else:
+          self.assertFalse('cbsdId' in response)
+
+      # Check response code matches the expected response code
+      self.assertIn(response['response']['responseCode'],config['expectedResponseCodes'][index])
+
+      if (response['response']['responseCode'] == 0) :
+        # Check if the 'availableChannel' parameter is present
+        self.assertTrue('availableChannel' in response)
+
+        # Check if the ruleApplied field is set to FCC_PART_96
+        for channel in response['availableChannel']:
+          self.assertEqual(channel['ruleApplied'], 'FCC_PART_96')
+
+        # Check that the GAA channel frequency ranges do not overlap with FR1cbsd ranges
+        for channel in response['availableChannel']:
+          if channel['channelType'] == 'GAA':
+            for fr1_range in config['fr1Cbsd'][index]:
+              # Check if the channel range is lower than the low freq of FR1cbsd range or if it is higher 
+              # than the high freq. If not then it is considered to be overlapping
+              self.assertTrue(((channel['frequencyRange']['highFrequency'] <= fr1_range['lowFrequency']) or
+                               (channel['frequencyRange']['lowFrequency'] >= fr1_range['highFrequency'])),
+                                "GAA channel overlaps with FR1cbsd")
+
+        # Check the PAL channels frequency ranges overlap with FR2cbsd
+        for channel in response['availableChannel']:
+          if channel['channelType'] == 'PAL':
+            for fr2_range in config['fr2Cbsd'][index]:
+              self.assertChannelsContainFrequencyRange([channel], fr2_range)
```

Comment:
Suppose fr2_range is [3550, 3560] and the SAS returns [ [3550, 3555], [3555, 3560] ]. I don't think it would pass this test case since `channel` will contain e.g. `[3550, 3555]`.

Can you pull out only the PAL channels, then do `self.assertChannelsContainFrequencyRange(pal_channels, fr2_range)`? Although that might cause trouble if there are multiple PAL channels for one CBSD :/

Reply:
Kate,

do you mean that the SAS would respond with 3550-3555 PAL and 3555-3560 PAL (i.e two separate 5 M PAL channels)?

I thought I heard/read somewhere that PAL channels are supposed to be 10M..... if that is not what you meant then disregard.


We rerun the latest code and ti's working fine

Reply:
Yes, I meant that the SAS could respond with two separate 5 MHz PAL channels. Masoud and I disagree on this exact point (5 MHz channels): he argues that since the PAL is 10 MHz, the Grants/SIQ responses for PAL should also be 10 MHz, whereas I argue that the PAL user should have at least as much flexibility as the GAA user (which *can* ask for a 5 MHz Grant). We haven't yet reached agreement on this issue, and I hope we can avoid needing to do so.

Anyway, we *do* agree in a different scenario: consider the case where there are two 10 MHz PPAs (both with this CBSD on their cluster list) on different channels, let's say 3550-3560 and 3600-3610. In that case, Masoud and I agree that the SAS should mark both as PAL channels. However, the current code would fail the SAS in that scenario. Do you agree?

I think the right logic here is "for each availableChannel marked PAL, there exists an fr2 which contains that frequency range." I believe that matches the test spec and should allow a correctly-implemented SAS to pass.

Reply:
so you are saying that before Janet's latest change if the config file was set such as fr2:

'fr2Cbsd': [
        [{'lowFrequency': 3550000000, 'highFrequency': 3560000000},
     {'lowFrequency': 36000000000, 'highFrequency': 36100000000}
],

it would not have worked?

but now with her latest change it'll work?

Reply:
Not quite. I'm saying that in the hypothetical scenario (i.e. a new config) where we have two PPAs on those channels (both with CBSD X on their cluster list) and we use fr2Cbsd as you wrote, the current code would fail a correctly-implemented SAS UUT which responded that both of those were PAL channels available to CBSD X.

Reply:
Added a utility to check that a given channel is in a list of frequency ranges and used it here instead. This PR is ready for re-review. Thanks!